app/wddr_main/main.c

```diff
@@ -227,40 +246,66 @@ void vAssertCalled( const char * const pcFileName, unsigned long ulLine )
 }
 
 /*-----------------------------------------------------------*/
-void handle_message(const Message_t *message)
+static BaseType_t handle_message(const Message_t *req, Message_t *rsp, phy_state_t *state)
 {
-    uint8_t freq_id;
-    Message_t resp_msg = {0, 0};
-
-    // Disable message interrupts while processing
-    interrupt_disable(MESSENGER_REQ_INTERRUPT_MSK |
-                      MESSENGER_ACK_INTERRUPT_MSK);
+    Message_t local_resp = {0x0, 0x0};
 
-    switch(message->id)
+    switch(req->id)
     {
         // Prep Request
         case MESSAGE_WDDR_FREQ_PREP_REQ:
+            uint8_t freq_id;
             // Extract Frequency ID
-            freq_id = GET_REG_FIELD(message->data, WDDR_FREQ_PREP_REQ__FREQ_ID);
+            freq_id = GET_REG_FIELD(req->data, WDDR_FREQ_PREP_REQ__FREQ_ID);
 
-            // Prepare for switch
-            if (firmware_phy_prep_switch(freq_id) == pdFAIL)
+            // Prepare for switch if phy is ready
+            if (*state != PHY_STATE_READY ||
+                firmware_phy_prep_switch(freq_id) == pdFAIL)
             {
                 // Mark as failure
-                resp_msg.data = UPDATE_REG_FIELD(0x0, WDDR_FREQ_PREP_RSP__STATUS, 0x1);
+                local_resp.data = UPDATE_REG_FIELD(0x0, WDDR_FREQ_PREP_RSP__STATUS, 0x1);
             }
 
-            resp_msg.id = MESSAGE_WDDR_FREQ_PREP_RESP;
-            resp_msg.data = UPDATE_REG_FIELD(resp_msg.data, WDDR_FREQ_PREP_RSP__FREQ_ID, freq_id);
-            wddr_messenger_send(&message_intf, &resp_msg);
+            // Prep was successful
+            local_resp.id = MESSAGE_WDDR_FREQ_PREP_RESP;
+            local_resp.data = UPDATE_REG_FIELD(local_resp.data, WDDR_FREQ_PREP_RSP__FREQ_ID, freq_id);
+            break;
+        case MESSAGE_GENERAL_MCU_BOOT_REQ:
+            // Boot PHY based on request
+            UBaseType_t ret  = firmware_phy_start(GET_REG_FIELD(req->data, WDDR_BOOT_REQ__CAL),
+                                                  GET_REG_FIELD(req->data, WDDR_BOOT_REQ__TRAIN));
+
+            // Return immediately if failure
+            // TODO: Is this the right choice? Should just tell host instead?
```

Comment:
Response has a status field for SUCCESS or FAILED.  Instead of returning, we should probably set that bit field in the response data.

Reply:
It does. But what are we supposed to do if boot fails? If boot fails, then something is wrong. Should the PHY keep handling requests? My gut says this should be treated as a fatal error which is how I coded it.